types/firestore.d.ts

```diff
@@ -91,6 +91,16 @@ declare namespace FirebaseFirestore {
      */
     toFirestore(modelObject: T): DocumentData;
 
+    /**
+     * Called by the Firestore SDK to convert a custom model object of
+     * Partial<T> into a plain Javascript object (suitable for writing directly
+     * to the Firestore database).
+     *
+     * This method must be defined in order to use set() with objects of type
+     * Partial<T>.
+     */
+    toFirestoreFromPartial?(modelObject: Partial<T>): DocumentData;
```

Comment:
Should this be `toFirestoreFromMerge`?

Even though you made this optional, this PR still smells like a breaking change to me since you added the validation that rejects calls if this is not set. I wonder if "merge" is a prominent enough part of our API that warrants that we should require that this is defined. What do you think? I am leaning 50% in favor of dropping the question mark.

Reply:
This would be a breaking change only for users using merge with converters. I don't see why anyone would be passing in `merge: true` with `<T>` objects, though I could see this breaking for existing developers using this with `mergeFields`.

However, wouldn't requiring this on the interface be a breaking change as well? Requiring `toFirestoreFromMerge()` would break developers using the converter without merge, and force developers who don't need to merge (ex: typical BulkWriter use case) to define a dummy method that they're not going to use. If I understand this correctly, the number of users that would be broken if we required this would be greater. Also the fact that we haven't received any feature requests for merge on Web could mean that there aren't that many developers using converters with merge. 

Considering that we're already making a breaking change with the change from `DocumentData` to `QueryDocumentSnapshot` in `fromFirestore()`, I'm leaning towards adding this as optional. Thoughts?

Reply:
I think either way this is a breaking change and we probably need to wait until later to roll this out to the Web SDK. What you have proposed is probably less breaking than requiring the interface.